Meme/Meme/NetworkClient.swift

```diff
@@ -8,49 +8,39 @@
 
 import Foundation
 
-protocol MemeClient {
-    func getMeme(from template: Template) -> Meme?
+protocol TemplateClient {
+    func getTemplates(_ data: Data) -> Result<[Template], RetrievalError>
 }
 
-class NetworkClient: MemeClient {
+class NetworkClient: TemplateClient {
 
-    fileprivate func getTemplates(_ data: Data?) -> [Template]? {
-        guard let data = data,
-            let jsonResponse = try? JSONSerialization.jsonObject(with: data, options: []) as? [String:String] else {
-                print("Error deserializing template data");
-                return nil
-        }
-        return jsonResponse.compactMap({ return Template.init(name: $0.0, templateURLString: $0.1)})
-    }
-
-    func requestTemplates(completion: @escaping ([Template]) -> Void)  {
+    func requestTemplates(completion: @escaping (Result<[Template], RetrievalError>) -> Void)  {
         let session = URLSession.shared
         let urlBuilder = MemeEndpointBuilder()
         let url = urlBuilder.endpointURL(.template)
         let request = URLRequest(url:url)
-        let task: URLSessionDataTask = session.dataTask(with: request) { (data, response, error) -> Void in
-            guard let templates = self.getTemplates(data) else { return }
-            completion(templates)
+        let task: URLSessionDataTask = session.dataTask(with: request) { [weak self] (data, response, error) -> Void in
```

Comment:
@vhart I added `[weak self]` since we will have retain cycles in the call to get templates. Prior to that it was not there, and I was able to do `completion(self.getTemplates(data))` without any local state since I updated the `getTemplates` method to return a result as per your suggestion.

However after adding `[weak self]` the call to `getTemplates` returns an optional `Result<[Template], RetrievalError>?` which is why I've had to unwrap it. Is that fine or is there a better way to do it?

Reply:
`(response as? HTTPURLResponse)?.apiError ` this part would need to be written but it's relatively straight forward as an extension of `HTTPURLResponse`
```swift 
extension HTTPURLResponse {
    var apiError: NetworkError? {
        switch statusCode {
        case 100...199:
            return .networkError
        case 200...299:
            return nil
        case 300...399:
            return nil
        case 400...499:
            return .clientError
        case 500...599:
            return .serverError
        default:
            return nil
        }
    }
}
```

Reply:
let me know what you think. Might be overkill

Reply:
I like it 👍